src/relay/backend/vm/deserializer.cc

```diff
@@ -0,0 +1,299 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file src/relay/backend/vm/deserializer.cc
+ * \brief Implementation of APIs to deserialize the serialized VM bytecode.
+ */
+
+#include "deserializer.h"
+
+#include <tvm/runtime/registry.h>
+#include <memory>
+#include <sstream>
+
+#include "serialize_util.h"
+
+namespace tvm {
+namespace relay {
+namespace vm {
+
+#define STREAM_CHECK(val, section)                                         \
+  CHECK(val) << "Invalid VM file format in the " << section << " section." \
+             << "\n";
+
+void Deserializer::Init(const std::string& code, const runtime::Module& lib) {
+  code_ = code;
+  vm_ = std::make_shared<VirtualMachine>();
+  vm_->lib = lib;
+}
+
+runtime::PackedFunc Deserializer::GetFunction(
+    const std::string& name,
+    const std::shared_ptr<ModuleNode>& sptr_to_self) {
+  if (name == "deserialize") {
+    return PackedFunc([sptr_to_self, this](TVMArgs args, TVMRetValue* rv) {
+      this->Deserialize();
+      *rv = runtime::Module(vm_);
+    });
+  } else {
+    LOG(FATAL) << "Unknown packed function: " << name;
+    return PackedFunc([sptr_to_self, name](TVMArgs args, TVMRetValue* rv) {});
+  }
+}
+
+void Deserializer::Deserialize() {
+  // Setup the stream.
+  dmlc::MemoryStringStream mss(&code_);
+  strm_ = &mss;
+
+  // Check header.
+  uint64_t header;
+  STREAM_CHECK(strm_->Read(&header), "header");
+  STREAM_CHECK(header == kTVMVMBytecodeMagic, "header");
+
+  // Check version.
+  std::string version;
+  STREAM_CHECK(strm_->Read(&version), "version");
+  STREAM_CHECK(version == TVM_VERSION, "version");
+
+  // Global section.
+  DeserializeGlobalSection();
+
+  // Constant section.
+  DeserializeConstantSection();
+
+  // Primitive names.
```

Comment:
Which primitive names are these?

Reply:
The TVM primitive ops that will be invoked by `InvokePacked`.